File: myulib-core/src/client/java/com/myudog/myulib/client/api/ui/component/TransformComponent.java
package com.myudog.myulib.client.api.ui.component;

import com.myudog.myulib.api.core.animation.AnimationTarget;
import com.myudog.myulib.api.core.ecs.IComponent;

public class TransformComponent implements IComponent {
    public float x;
    public float y;
    public float width;
    public float height;
    public float scaleX = 1.0f;
    public float scaleY = 1.0f;
    public float rotation;
    public float opacity = 1.0f;

    public AnimationTarget<Float> xTarget() {
        return value -> this.x = value;
    }

    public AnimationTarget<Float> yTarget() {
        return value -> this.y = value;
    }

    public AnimationTarget<Float> widthTarget() {
        return value -> this.width = value;
    }

    public AnimationTarget<Float> heightTarget() {
        return value -> this.height = value;
    }

    public AnimationTarget<Float> scaleXTarget() {
        return value -> this.scaleX = value;
    }

    public AnimationTarget<Float> scaleYTarget() {
        return value -> this.scaleY = value;
    }

    public AnimationTarget<Float> rotationTarget() {
        return value -> this.rotation = value;
    }

    public AnimationTarget<Float> opacityTarget() {
        return value -> this.opacity = value;
    }
}

File: src/client/kotlin/com/myudog/myulib/client/api/ui/component/TransformComponent.kt
package com.myudog.myulib.client.api.ui.component

import com.myudog.myulib.api.ecs.Component
import com.myudog.myulib.client.api.ui.Anchor
import com.myudog.myulib.client.api.ui.BoxValues
import com.myudog.myulib.client.api.ui.SizeUnit

/**
 * [API] 變換組件：定義 UI 元件的佈局規則。
 * 這裡不儲存最終的像素座標，而是儲存「如何計算座標」的意圖。
 */
data class TransformComponent(
    var anchor: Anchor = Anchor.TOP_LEFT,
    var width: SizeUnit = SizeUnit.Fixed(100f),
    var height: SizeUnit = SizeUnit.Fixed(20f),
    var offsetX: Float = 0f,
    var offsetY: Float = 0f,
    // 內距與外距預留
    var padding: BoxValues = BoxValues.ZERO,
    var margin: BoxValues = BoxValues.ZERO
) : Component
